File: sample/SampleIsolatedFunctions.V4/TestFunction.cs
// <copyright file="TestFunction.cs" company="DarkLoop" author="Arturo Martinez">
//  Copyright (c) DarkLoop. All rights reserved.
// </copyright>

using System.Text;
using DarkLoop.Azure.Functions.Authorization;
using Microsoft.AspNetCore.Authentication;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace SampleIsolatedFunctions.V4
{
    [FunctionAuthorize(AuthenticationSchemes = "Bearer")]
    public class TestFunction
    {
        private readonly ILogger<TestFunction> _logger;

        public TestFunction(ILogger<TestFunction> logger)
        {
            _logger = logger;
        }

        [Function("TestFunction")]
        [Authorize(Roles = "admin")]
        public async Task<IActionResult> Run([HttpTrigger("get", "post")] HttpRequest req)
        {
            _logger.LogInformation("C# HTTP trigger function processed a request.");

            var provider = req.HttpContext.RequestServices;
            var schProvider = provider.GetService<IAuthenticationSchemeProvider>();

            var sb = new StringBuilder();
            sb.AppendLine("Authentication schemes:");

            if (schProvider is not null)
            {
                foreach (var scheme in await schProvider.GetAllSchemesAsync())
                    sb.AppendLine($"  {scheme.Name} -> {scheme.HandlerType}");
            }

            sb.AppendLine();
            sb.AppendLine($"User:");
            sb.AppendLine($"  Name  -> {req.HttpContext.User.Identity!.Name}");
            sb.AppendLine($"  Email -> {req.HttpContext.User.FindFirst("email")?.Value}");

            return new OkObjectResult(sb.ToString());
        }
    }
}

File: sample/SampleIsolatedFunctionsFSharp.V4/TestFunction.fs
namespace SampleIsolatedFunctionsFSharp.V4

open System.Text
open DarkLoop.Azure.Functions.Authorization
open Microsoft.AspNetCore.Authentication
open Microsoft.AspNetCore.Authorization
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Mvc
open Microsoft.Azure.Functions.Worker
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Logging


[<FunctionAuthorize(AuthenticationSchemes = "FunctionsBearer")>]
type TestFunction(logger:ILogger<TestFunction>) =
    let _logger = logger
        
    [<Function("TestFunction")>]
    [<Authorize(Roles = "admin")>]
    member _.Run([<HttpTrigger("get", "post")>] req:HttpRequest) =
                task {
                        _logger.LogInformation("F# HTTP trigger function processed a request.")

                        let provider = req.HttpContext.RequestServices
                        let schProvider = provider.GetService<IAuthenticationSchemeProvider>()

                        let sb = new StringBuilder()
                        sb.AppendLine("Authentication schemes:") |> ignore

                        if (schProvider <> null) then             
                            let! allScheme = schProvider.GetAllSchemesAsync()
                            for scheme in allScheme do
                                sb.AppendLine($"  {scheme.Name} -> {scheme.HandlerType}") |> ignore
             

                        sb.AppendLine()|> ignore
                        sb.AppendLine($"User:")|> ignore
                        sb.AppendLine($"  Name  -> {req.HttpContext.User.Identity.Name}")|> ignore
                        let email = req.HttpContext.User.FindFirst("email")|> Option.ofObj|>Option.map _.Value
                        sb.AppendLine($"  Email -> {email}")|> ignore

                        return OkObjectResult(sb.ToString())
            }